### ISSUE TITLE
enable syncing to a prebuilt image with no docker push

### ISSUE THREAD

**Issue by landism** [windmilleng/tilt#2582]
sample use case: I have a docker image that's built by CI, and is close enough to the code I want to run. I want to `tilt up`, have that image deployed, and then be able to live_update to it.

This is soooort of achievable at the moment by specifying a `docker_build` whose Dockerfile is just a one-line "FROM base-image", but that comes with the potentially major caveat that I have to pull that image to my laptop and then push it back up, which might be annoyingly expensive.

**Comment by digger18** [comment]
Another possible hack is to use custom_build and instead docker build use docker tag.

**Comment by maiamcc** [comment]
Note that this isn't quiiite as simple as "for a deployed image that we didn't build, apply existing live_update logic: whenever a `sync`'d file changes on disk, copy it up to the running container and execute any `run` steps."

we'll need a way of syncing up current code state if e.g. the pod goes down and we lose in-memory changes, which will require us to be able to do a one-off run of ALL live_update steps, i.e. sync ALL sync'd directories, run ALL run steps.

I think this issues dovetails with a feature requested by @drubin, right? (Or is there another ticket for that somewhere?)

**Comment by drubin** [comment]
Yes it is something I am interested in. Been following this waiting :)

**Comment by arrrnas** [comment]
We have the same use-case as OP. I've implemented this as a `live_update` step called `initial_sync`:
```python
  live_update=[
      initial_sync(ignore=['node_modules', '.git', '**/test']),
      sync('./src', '/app/src'),
      run('npm install', trigger=['package.json']),
  ]
```

Our setup: We use `k8s_custom_deploy` to deploy pods using pre-built images (for performance). Images are built from commits on the main branch, so they differ from local files unless you're on the exact same commit with no changes.           

How it works:                                                                                                                                                                                                                                    
  - On container start or restart, all files from sync paths are collected and transferred to the container
  - run steps execute based on whether their trigger files are in the synced set — steps without triggers always run, steps with triggers that don't match are skipped (to avoid failures from missing files)
  - ignore accepts dockerignore-style glob patterns `(**/test, node_modules, *.log)`
  - An optional dockerignore parameter points to a directory with a .dockerignore to reuse
                              
Performance: For larger projects (~20k files, ~90MB) sync takes under 20s to a remote EKS cluster. I initially implemented gzip compression but dropped it — the CPU overhead of compressing many small files was greater than the network transfer savings.

vs file_sync_only (tilt-dev/tilt-extensions#77): That extension only syncs once and lacks control over what gets synced. initial_sync runs on every container start/restart, supports ignore patterns, and integrates with the existing run trigger system.                                                                                                                                                                                                                                

Draft PR: https://github.com/tilt-dev/tilt/pull/6737
Draft documentation PR: https://github.com/tilt-dev/tilt.build/pull/1197

This also partially addresses #3961 and #4776.
